File: Implementierung/Backend/src/main/java/edu/kit/informatik/dto/mapper/courses/CourseMapper.java
package edu.kit.informatik.dto.mapper.courses;

import edu.kit.informatik.dto.mapper.IModelDtoMapper;
import edu.kit.informatik.dto.userdata.courses.CourseDto;
import edu.kit.informatik.model.User;
import edu.kit.informatik.model.userdata.courses.Course;
import edu.kit.informatik.model.userdata.courses.SeatArrangement;
import edu.kit.informatik.model.userdata.courses.Session;
import edu.kit.informatik.model.userdata.interactions.Participant;
import edu.kit.informatik.repositories.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

/**
 * {@link IModelDtoMapper} für {@link Course} and {@link CourseDto}
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Service
public class CourseMapper implements IModelDtoMapper<Course, CourseDto> {

    private final UserRepository userRepository;
    private final ParticipantRepository participantRepository;
    private final SessionRepository sessionRepository;
    private final SeatArrangementRepository seatArrangementRepository;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     * @param userRepository {@link UserRepository}
     * @param participantRepository {@link ParticipantRepository}
     * @param sessionRepository {@link SessionRepository}
     * @param seatArrangementRepository {@link SeatArrangementRepository}
     */
    @Autowired
    public CourseMapper(UserRepository userRepository,
                        ParticipantRepository participantRepository, SessionRepository sessionRepository,
                        SeatArrangementRepository seatArrangementRepository) {
        this.userRepository = userRepository;
        this.participantRepository = participantRepository;
        this.sessionRepository = sessionRepository;
        this.seatArrangementRepository = seatArrangementRepository;
    }

    @Override
    public CourseDto modelToDto(Course course) {
        List<String> sessionIds = new ArrayList<>();
        List<String> participantIds = new ArrayList<>();
        List<String> seatArrangementIds = new ArrayList<>();

        course.getSessions().forEach(session -> sessionIds.add(session.getId()));
        course.getParticipants().forEach(participant -> participantIds.add(participant.getId()));
        course.getSeatArrangements().forEach(seatArrangement -> seatArrangementIds.add(seatArrangement.getId()));

        return new CourseDto(
                course.getId(),
                course.getUser().getId(),
                course.getName(),
                course.getSubject(),
                sessionIds,
                participantIds,
                seatArrangementIds,
                course.getCreatedAt(),
                course.getUpdatedAt()
        );
    }

    @Override
    public List<CourseDto> modelToDto(List<Course> courses) {
        List<CourseDto> courseDtos = new ArrayList<>();
        courses.forEach(course -> courseDtos.add(modelToDto(course)));

        return courseDtos;
    }

    @Override
    public Course dtoToModel(CourseDto courseDto) {

        User user = userRepository.findUserById(courseDto.getUserId()).orElse(null);

        List<Participant> participants = new ArrayList<>();
        List<Session> sessions = new ArrayList<>();
        List<SeatArrangement> seatArrangements = new ArrayList<>();

        if (courseDto.getParticipantIds() != null) {
            courseDto.getParticipantIds().forEach(id ->
                    participantRepository.findParticipantById(id).ifPresent(participants::add));
        }
        if (courseDto.getSessionIds() != null) {
            courseDto.getSessionIds().forEach(id ->
                    sessionRepository.findSessionById(id).ifPresent(sessions::add));
        }
        if (courseDto.getSeatArrangementIds() != null) {
            courseDto.getSeatArrangementIds().forEach(id ->
                    seatArrangementRepository.findSeatArrangementById(id).ifPresent(seatArrangements::add));
        }

        Timestamp updatedAt;

        if (courseDto.getUpdatedAt() == null) {
            updatedAt = courseDto.getCreatedAt();
        } else {
            updatedAt = courseDto.getUpdatedAt();
        }

        Course course = new Course(user, courseDto.getName(), courseDto.getSubject(),
                                        courseDto.getCreatedAt(), updatedAt);

        course.setParticipants(participants);
        course.setSessions(sessions);
        course.setSeatArrangements(seatArrangements);

        return course;
    }

    @Override
    public List<Course> dtoToModel(List<CourseDto> courseDtos) {
        List<Course> courses = new ArrayList<>();
        courseDtos.forEach(courseDto -> courses.add(dtoToModel(courseDto)));

        return courses;
    }
}

File: Implementierung/frontend/paschi/src/dto/mapper/courses/CourseMapper.ts
import {IModelDtoMapper} from "@/dto/mapper/IModelDtoMapper";
import {Course} from "@/model/userdata/courses/Course";
import {CourseDto} from "@/dto/userdata/courses/CourseDto";
import {Session} from "@/model/userdata/courses/Session";
import {Participant} from "@/model/userdata/interactions/Participant";
import {SeatArrangement} from "@/model/userdata/courses/SeatArrangement";
import {UserController} from "@/controller/UserController";
import {useCourseStore} from "@/store/CourseStore";
import {useStudentStore} from "@/store/StudentStore";
import {ParticipantService} from "@/service/ParticipantService";
import {useSessionStore} from "@/store/SessionStore";
import {SessionService} from "@/service/SessionService";
import {SeatArrangementService} from "@/service/SeatArrangementService";
import {useSeatArrangementStore} from "@/store/SeatArrangementStore";

export class CourseMapper implements IModelDtoMapper<Course, CourseDto> {

  private static mapper: CourseMapper = new CourseMapper();

  private constructor() {
  }

  static getMapper(): IModelDtoMapper<Course, CourseDto> {
    return CourseMapper.mapper;
  }

  modelToDto(course: Course): CourseDto {
    const sessionIds: string[] = [];
    const participantIds: string[] = [];
    const arrangementIds: string[] = [];

    course.sessions.forEach((session: Session) => sessionIds.push(session.getId));
    course.participants.forEach((participant: Participant) => participantIds.push(participant.getId));
    course.seatArrangements.forEach((arrangement: SeatArrangement) => arrangementIds.push(arrangement.getId));

    return new CourseDto(
      course.getId,
      course.user.getId,
      course.createdAt,
      course.updatedAt,
      course.name,
      course.subject,
      sessionIds,
      participantIds,
      arrangementIds
    );
  }

  async dtoToModel(courseDto: CourseDto): Promise<Course> {
    const userController = UserController.getUserController();

    let course = useCourseStore().getCourse(courseDto.id);
    if (course == undefined) {
      course = new Course(
        courseDto.id,
        0,
        userController.getUser(),
        courseDto.name,
        courseDto.subject
      );
      course.createdAt = courseDto.createdAt;
      course.updatedAt = courseDto.updatedAt;
      useCourseStore().addCourse(course);
    } else if (course.updatedAt === courseDto.updatedAt) {
      return course;
    } else {
      course.name = courseDto.name;
      course.subject = courseDto.subject;
    }

    const sessionService = SessionService.getService();
    const participantService = ParticipantService.getService();
    const arrangementService = SeatArrangementService.getService();

    const sessions: Session[] = [];
    const participants: Participant[] = [];
    const arrangements: SeatArrangement[] = [];

    for (const id of courseDto.participantIds) {
      let participant = useStudentStore().getStudent(id);
      if (participant == undefined) {
        participant = await participantService.getById(id);
      }
      if (participant !== undefined) {
        participants.push(participant);
      }
    }

    for (const id of courseDto.sessionIds) {
      let session = useSessionStore().getSession(id);
      if (session == undefined) {
        session = await sessionService.getById(id);
      }
      if (session !== undefined) {
        sessions.push(session);
      }
    }

    for (const id of courseDto.seatArrangementIds) {
      let arrangement = useSeatArrangementStore().getSeatArrangement(id);
      if (arrangement == undefined) {
        arrangement = await arrangementService.getById(id);
      }
      if (arrangement !== undefined) {
        arrangements.push(arrangement);
      }
    }

    course.participants = participants;
    course.sessions = sessions;
    course.seatArrangements = arrangements;

    return course;
  }
}
